libazureinit/src/config.rs

```diff
@@ -0,0 +1,857 @@
+    pub ssh: Ssh,
+    pub hostname_provisioners: HostnameProvisioners,
+    pub user_provisioners: UserProvisioners,
+    pub password_provisioners: PasswordProvisioners,
+    pub imds: Imds,
+    pub provisioning_media: ProvisioningMedia,
+    pub azure_proxy_agent: AzureProxyAgent,
+    pub wireserver: Wireserver,
+    pub telemetry: Telemetry,
+}
+
+/// Implements `Display` for `Config`, formatting it as a readable TOML string.
+///
+/// Uses `toml::to_string_pretty` to serialize the configuration. If serialization fails,
+/// a fallback message is displayed..
+impl fmt::Display for Config {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            toml::to_string_pretty(self)
+                .unwrap_or_else(|_| "Unable to serialize config.".to_string())
+        )
+    }
+}
+
+/// Loads the configuration for `azure-init`.
+///
+/// This method uses the `Figment` library to load configuration from the following sources,
+/// in order of priority:
+///
+/// 1. **Defaults**: Base configuration from `Config::default()`.
+/// 2. **Main File**: `azure-init.toml`, if present.
+/// 3. **Directory Files**: `.toml` files in `azure-init.d`, sorted lexicographically.
+/// 4. **CLI Overrides**: A file or directory specified via the CLI.
+///
+/// Later sources override earlier ones in case of conflicts.
+impl Config {
+    pub fn load(path: Option<PathBuf>) -> Result<Config, Error> {
+        let mut figment =
+            Figment::new().merge(Serialized::defaults(Config::default()));
+
+        if PathBuf::from("azure-init.toml").exists() {
+            tracing::info!("Loading base configuration file: azure-init.toml");
+            figment = figment.merge(Toml::file("azure-init.toml"));
+        } else {
+            tracing::warn!("Base configuration file azure-init.toml not found, using defaults.");
+        }
+
+        figment = Self::merge_toml_directory(
+            &figment,
+            PathBuf::from("azure-init.toml.d"),
+        )?;
+
+        if let Some(cli_path) = path {
+            if cli_path.is_dir() {
+                figment = Self::merge_toml_directory(&figment, cli_path)?;
+            } else {
+                tracing::info!(
+                    "Merging configuration file from CLI: {:?}",
+                    cli_path
+                );
+                figment = figment.merge(Toml::file(cli_path));
+            }
+        }
+
+        figment.extract::<Config>().map_err(|e| {
+            tracing::error!("Failed to extract configuration: {:?}", e);
+            Error::Io(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                format!("Configuration error: {:?}", e),
+            ))
+        })
+    }
+
+    /// Helper function to merge `.toml` files from a directory into the Figment configuration.
+    fn merge_toml_directory(
+        figment: &Figment,
+        dir_path: PathBuf,
+    ) -> Result<Figment, Error> {
```

Comment:
An alternative pattern you can use here is:

```Rust
fn merge_toml_directory(
    mut figment: Figment,
    dir_path: PathBuf,
) -> Result<Figment, Error> {
```

This gives you ownership of the incoming Figment object, which you then modify before passing ownership _back_ to the caller. With this approach you don't need to do

```Rust
            let mut updated_figment = figment.clone();
            for path_entry in entries {
                tracing::info!("Merging configuration file: {:?}", path_entry);
                updated_figment = updated_figment.merge(Toml::file(path_entry));
            }
            Ok(updated_figment)
```

and can instead do

```Rust
            for path_entry in entries {
                tracing::info!("Merging configuration file: {:?}", path_entry);
                figment = figment.merge(Toml::file(path_entry));
            }
            Ok(figment)
```

It allows you to avoid an unnecessary clone. This is obviously not a performance-sensitive bit of code so avoiding a clone is not something we care much about, but it's a pattern you'll see lots of places (figment itself is doing it with the `merge` function) and it's nice to be aware of.